### ISSUE TITLE
Serializing SymEngine objects

### ISSUE THREAD

**Issue by bjodah** [symengine/symengine#1394]
There are quite a few applications where serialization of SymEngine objects would be useful.
One is pickling support in Python (as requested in https://github.com/symengine/symengine.py/issues/205).
Another would be e.g. efficient dump/load to disk of large expressions without having to parse expressions.

I am not well versed in the plethora of C++ serialization libraries, I know of boost::serialize and [cereal](https://github.com/USCiLab/cereal) but I haven't written anything against them myself. Are those preferable to e.g. implementing `loads`/`dumps` (de-)serialization functions on e.g. `Basic`, `DenseMatrix`, etc. (but what about `vec_basic` then?).

Or is this a terrible idea altogether?

**Comment by isuruf** [comment]
@bluescarni, @certik, any recommendations?

**Comment by bluescarni** [comment]
IMO it largely depends on the use cases you have in mind for serialization.

Libraries like cereal and boost serialization make it quite easy to implement the serialization logic. They support out of the box the STL containers, and they map particularly well to a language like C++ in which you frequently have user-defined types nested into other user-defined types. They can also perform very well, and they understand specific C++ constructs like inheritance, virtual classes, etc. Their 2 main downsides are that 1) they serialize into some custom format, which is not much usable as a data exchange format with other libraries/programs 2) they offer few guarantees of forward compatibility (e.g., if you serialize an object with Boost 1.66 it's not guaranteed to be deserializable with Boost 1.65).

I have some experience with msgpack as well (https://github.com/msgpack/msgpack-c). msgpack is something like a JSON binary format: it retains JSON's simple model but with much higher performance and leaner archive size. If your data structures map well to JSON semantics (e.g., arrays, dicts, strings, etc.), it can be a good choice: the data format is stable and well documented, so you can use it as a way to transfer data from symengine into some other application that speaks JSON/msgpack, and you can also use it as a long-term storage format. The main downside of this solution is that JSON represents natively only a subset of C++ objects: if you need to serialize, for instance, derived virtual classes through a pointer to base, you'll have to come up with some way of mapping this into JSON on your own.

Finally, there are solutions like protobuf/thrift/flatbuffers/etc. I have never used these myself, when I looked into them I found them unconvincing for my use cases because a) they require some external tooling to work b) they did not seem to support natively various C++ features I was interested in. However, they seem to guarantee a high degree of back/forward compatibility and they seem also to perform very well.

**Comment by bjodah** [comment]
Thank you @bluescarni for your insights.

So I've come full circle back to this issue in a project of mine in which I'm using [cereal](https://uscilab.github.io/cereal/). Writing a short `serialize` method or `load`/`save` pair allows me to store objects on disk or send them over the network. So far so good.

Some of my structs have e.g. a `std::vector<SymEngine::RCP<const SymEngine::Basic>>` member. In this case I know that those Basic instances are scalar expressions, and should therefore survive a `__str__` -> `SymEngine::parse(...)` round-trip. So I write a save/load pair of functions templates in the `cereal` namespace (per cereals docs) for our RCP<const Basic>, but I cannot seem to get it to work. The delights of C++ compilation error messages aren't exactly helping either (with tons of enable_if and SFINAE magic in cereal code base---out of necessity of course) 

Here's a minimal example:
```C++
#include <symengine/basic.h>
#include <symengine/parser.h>

#include <cereal/archives/binary.hpp>

namespace se = SymEngine;
using se_basic = se::RCP<const se::Basic>;


namespace cereal {
    template<typename Archive>
    void save(Archive& ar, const se_basic& b) {
        ar(b->__str__());
    }

    template <typename Archive> void load(Archive &ar, se_basic &b) {
        std::string srepr;
        ar(srepr);
        b = se::parse(srepr);
    }
} // namespace cereal

int main() {
    se_basic x = se::parse("x + 1");
    std::ostringstream oss;
    cereal::BinaryOutputArchive boa{oss};
    boa(x);
    // logic to save to string and load from string omitted since it already fails to compile
    return 0;
}
```
<details>
<summary>Output from compilation</summary>
<pre>
In file included from /usr/include/cereal/archives/binary.hpp:32,
                 from /home/bjorn/doc/src/cpp/cereal/serial3_syme.cpp:5:
/usr/include/cereal/cereal.hpp: In instantiation of ‘ArchiveType& cereal::OutputArchive<ArchiveType, Flags>::processImpl(const T&) [with T = std::__cxx11::basic_string<char>; typename cereal::traits::detail::EnableIfHelper<(cereal::traits::has_invalid_output_versioning<T, ArchiveType>::value || ((! cereal::traits::is_output_serializable<T, ArchiveType>::value) && ((!(Flags & cereal::AllowEmptyClassElision)) || ((Flags & cereal::AllowEmptyClassElision) && (! std::is_empty<T>::value)))))>::type <anonymous> = (cereal::traits::detail::sfinae)0; ArchiveType = cereal::BinaryOutputArchive; unsigned int Flags = 1]’:
/usr/include/cereal/cereal.hpp:417:26:   required from ‘void cereal::OutputArchive<ArchiveType, Flags>::process(T&&) [with T = std::__cxx11::basic_string<char>; ArchiveType = cereal::BinaryOutputArchive; unsigned int Flags = 1]’
/usr/include/cereal/cereal.hpp:311:22:   required from ‘ArchiveType& cereal::OutputArchive<ArchiveType, Flags>::operator()(Types&& ...) [with Types = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; ArchiveType = cereal::BinaryOutputArchive; unsigned int Flags = 1]’
/home/bjorn/doc/src/cpp/cereal/serial3_syme.cpp:14:11:   required from ‘void cereal::save(Archive&, const se_basic&) [with Archive = cereal::BinaryOutputArchive; se_basic = Teuchos::RCP<const SymEngine::Basic>]’
/usr/include/cereal/cereal.hpp:505:34:   required from ‘ArchiveType& cereal::OutputArchive<ArchiveType, Flags>::processImpl(const T&) [with T = Teuchos::RCP<const SymEngine::Basic>; typename cereal::traits::detail::EnableIfHelper<cereal::traits::has_non_member_save<T, ArchiveType>::value, (! cereal::traits::has_invalid_output_versioning<T, ArchiveType>::value), (cereal::traits::is_output_serializable<T, ArchiveType>::value && (cereal::traits::is_specialized_non_member_save<T, ArchiveType>::value || (! cereal::traits::is_specialized<T, ArchiveType>::value)))>::type <anonymous> = (cereal::traits::detail::sfinae)0; ArchiveType = cereal::BinaryOutputArchive; unsigned int Flags = 1]’
/usr/include/cereal/cereal.hpp:417:26:   required from ‘void cereal::OutputArchive<ArchiveType, Flags>::process(T&&) [with T = Teuchos::RCP<const SymEngine::Basic>&; ArchiveType = cereal::BinaryOutputArchive; unsigned int Flags = 1]’
/usr/include/cereal/cereal.hpp:311:22:   required from ‘ArchiveType& cereal::OutputArchive<ArchiveType, Flags>::operator()(Types&& ...) [with Types = {Teuchos::RCP<const SymEngine::Basic>&}; ArchiveType = cereal::BinaryOutputArchive; unsigned int Flags = 1]’
/home/bjorn/doc/src/cpp/cereal/serial3_syme.cpp:28:10:   required from here
/usr/include/cereal/cereal.hpp:543:87: error: static assertion failed: cereal could not find any output serialization functions for the provided type and archive combination. 

 Types must either have a serialize function, load/save pair, or load_minimal/save_minimal pair (you may not mix these). 
 Serialize functions generally have the following signature: 

 template<class Archive> 
   void serialize(Archive & ar) 
   { 
     ar( member1, member2, member3 ); 
   } 

 
  543 |         static_assert(traits::detail::count_output_serializers<T, ArchiveType>::value != 0,
      |                                                                                 ~~~~~~^~~~
gmake[2]: *** [CMakeFiles/serial3_syme.dir/build.make:63: CMakeFiles/serial3_syme.dir/serial3_syme.cpp.o] Error 1
gmake[2]: Leaving directory '/home/bjorn/doc/src/cpp/cereal/build'
gmake[1]: *** [CMakeFiles/Makefile2:105: CMakeFiles/serial3_syme.dir/all] Error 2
gmake[1]: Leaving directory '/home/bjorn/doc/src/cpp/cereal/build'
gmake: *** [Makefile:84: all] Error 2
</pre>
</details>

Any chance someone has gotten this to work in the mean time? Or perhaps is able to spot the error magically ^^ ?

Here's a gist with this file, a working example and a CMakeLists.txt: https://gist.github.com/bjodah/85fbf793da29a405e71ee382746db94d (in case someone has empty cycles and feels merciful).

**Comment by certik** [comment]
I assume Cereal requires some properties that SymEngine RCP does not satisfy, but the error message is quite cryptic.

I think SymEngine should have some native serialization that is high performing, both load and save. There are multiple ways, but the most straightforward is: Each SymEngine class saves its type as an integer, and after that a custom serialization of that class. So a `Symbol` would save in this order:

* an integer type
* length of the string
* a string (the name of the symbol)

An `Add` and `Mul` would save:

* an integer type
* how many operands
* each operand saved using its serialization routine

`Pow` would have:

* an integer type
* base
* exponent

Etc. This is the most straightforward approach, it should be fast to save and fast to load. The possible disadvantage is that it might be quite long for long expressions, since every term has to save the `Symbol` from scratch. Possible optimizations are to have a symbol table and just reference this table etc., but that gets more complicated to implement. We can have multiple serialization routines based on the serialization algorithm. But the above should go a long way for most use cases.

**Comment by bjodah** [comment]
Yes, thanks, I too guess that's whats happening. I was trying to figure out how they've implemented the same functionality for std::shared_ptr, but it's trickier than I anticipated. I'll work around this issue by other means.

Regarding the serialization. Yes, something along those lines would definitely be useful. Reflection will hopefully be part of C++ by C++23, that would probably make it easier to implement. Right now, using cereal, I find myself repeating the the names of the members over and over...

**Comment by certik** [comment]
In LFortran I generate all the classes from ASDL automatically: https://gitlab.com/lfortran/lfortran/-/blob/63fdfeef69e766da00234d3ddd684fa2454e0bda/grammar/AST.asdl, and so adding a serializer would be just a simple extension of the generator script. I think the exact same approach would work for SymEngine also.

**Comment by bjodah** [comment]
Interesting references @certik. Thank you for sharing, and your LFortran project looks impressive.

**Comment by isuruf** [comment]
@bjodah, RCP is similar to `boost::intrusive_ptr`, so you can have a look at how that is implemented in https://github.com/USCiLab/cereal/pull/282/files

**Comment by bluescarni** [comment]
> Here's a gist with this file, a working example and a CMakeLists.txt: https://gist.github.com/bjodah/85fbf793da29a405e71ee382746db94d (in case someone has empty cycles and feels merciful).

I believe you are missing ``#include <cereal/types/string.hpp>`` - the error is not in the serialization of the symengine bits, it happens when you try to serialize the string representation returned by ``b->__str__()``.

**Comment by bluescarni** [comment]
> I assume Cereal requires some properties that SymEngine RCP does not satisfy, but the error message is quite cryptic.
> 
> I think SymEngine should have some native serialization that is high performing, both load and save. There are multiple ways, but the most straightforward is: Each SymEngine class saves its type as an integer, and after that a custom serialization of that class. So a `Symbol` would save in this order:
> 
>     * an integer type
> 
>     * length of the string
> 
>     * a string (the name of the symbol)
> 
> 
> An `Add` and `Mul` would save:
> 
>     * an integer type
> 
>     * how many operands
> 
>     * each operand saved using its serialization routine
> 
> 
> `Pow` would have:
> 
>     * an integer type
> 
>     * base
> 
>     * exponent
> 
> 
> Etc. This is the most straightforward approach, it should be fast to save and fast to load. The possible disadvantage is that it might be quite long for long expressions, since every term has to save the `Symbol` from scratch. Possible optimizations are to have a symbol table and just reference this table etc., but that gets more complicated to implement. We can have multiple serialization routines based on the serialization algorithm. But the above should go a long way for most use cases.

This is just my opinion of course, but I'd strongly advise against doing this. Serialization is a well understood problem and, apart from the clunkiness arising from the lack of true introspection capabilities in the language, there are robust, battle-tested libraries that take care of **many** tricky details, some of which you mention (e.g., avoiding storing multiple times the same pointer/object, serialization via base classes, smart pointers, etc.). I personally am partial to Boost.Serialization and I'd prefer it over cereal, but there are many other options as well.

I would look into a home-grown solution only after having verified that the existing ones don't fit the bill for some reason...

**Comment by bjodah** [comment]
@bluescarni I could have sworn I had added the string header---how embarrassing, you're of course right! 😅 
A simple `#include <cereal/types/string.hpp>` and all is fine (if only c++ had better error messages...). Thank you 🙏

Just curious about your experiences with cereal vs. boost, what makes you prefer boost.serialization? I've noticed that cereal doesn't see much development lately, boost is perhaps a better bet just because of that.

**Comment by certik** [comment]
@bluescarni you are right. What worries me with the Boost link is that under "Our goals for such a system are:" they do not mention performance / speed. The advanage of the above "simple" approach is that you dump the memory pretty much (no custom logic), so it's hard to beat that in terms of speed, and it is straightforward to implement --- I can do it for LFortran first to show the advantages, since I can auto-generate it.

**Comment by bluescarni** [comment]
> Just curious about your experiences with cereal vs. boost, what makes you prefer boost.serialization? I've noticed that cereal doesn't see much development lately, boost is perhaps a better bet just because of that.

It's a bit convoluted, but I'll try to explain.

I was using cereal in projects which required serialization of derived classes via pointers to the base classes. In order to accomplish this, serialization libraries (both cereal and Boost) need to keep around and manipulate global objects which manage the correspondence between the base and derived classes at runtime.

Because cereal is header-only, these objects are declared in the headers and they are instantiated on-demand. This works fine as long as your project consists of one executable or one shared library - even if multiple copies of these global objects are produced in multiple translation units, only a single copy will be kept when the linker produces the final object.

The problems start when you are using cereal in multiple projects and you are doing things like ``dlopen()``ing different libraries which are using cereal internally. This was the case in a couple of projects which were thin Python wrappers around C++ libraries, both using cereal as a serialization backend. In this specific setup, it would happen that we ended up with two different copies of these internal global cereal objects, which would create really hard-to-debug crashes at runtime.

Boost.Serialization, on the other hand, is not header-only (for this very reason I suspect) and it keeps all the mutable global data isolated in its own shared library, and thus there is never the risk of multiple copies of the same data structures stepping on each other's toes at runtime.

This was admittedly a bit of a corner case, but it made me appreciate the insight that went into designing the architecture of Boost.Serialization (and as an aside it also taught me that if you ever need global mutable state in your library, then it **has to** live in a compiled part of the library and not in the headers, however tempting that might be).

**Comment by bluescarni** [comment]
> @bluescarni you are right. What worries me with the Boost link is that under "Our goals for such a system are:" they do not mention performance / speed. The advanage of the above "simple" approach is that you dump the memory pretty much (no custom logic), so it's hard to beat that in terms of speed, and it is straightforward to implement --- I can do it for LFortran first to show the advantages, since I can auto-generate it.

I can sympathize with the performance argument, and yes, a memory dump will be hard to beat in terms of speed.

On the other hand, I have a bit of a hard time imagining scenarios in which serialization performance is going to be a bottleneck for a symbolic expression library, but maybe I am not trying hard enough :)

**Comment by certik** [comment]
> On the other hand, I have a bit of a hard time imagining scenarios in which serialization performance is going to be a bottleneck for a symbolic expression library, but maybe I am not trying hard enough :)

Yes, so if we do not need performance, then we have more options. I always assume with SymEngine that we want the most efficient implementation possible.

**Comment by isuruf** [comment]
https://github.com/isuruf/symengine/tree/cereal should have an initial implementation for a few objects. The implementation for Integer could be done more efficiently, but should be good for now. Needs tests. @bjodah, would you like to help finish it?

> Yes, so if we do not need performance, then we have more options. I always assume with SymEngine that we want the most efficient implementation possible.

True, but having a "maybe inefficient but still faster than sympy" option would be useful to have

**Comment by bjodah** [comment]
@bluescarni, thank you for writing this up, most valuable (I can only begin to imagine the horrors of debugging that ^^).
I could easily imagine that there will be libraries using symengine (maybe linking statically with hidden visibility), let's say we have two such libraries, libfoo & libbar, then there are perhaps two python wrappers for those: pyfoo and pybar, both relying on SymEngine's possible future serialization capabilities to support Python's pickle interface. If someone imports both pyfoo and pybar, and we're using cereal, am I correct that this would lead to the kind of situation you just described?

@isuruf, impressively fast as always :+1: . Sure, I'd love to help 🙂, I just forked your branch and compiled it. Will start working on the tests. I'll open a PR against your repo?

**Comment by isuruf** [comment]
Created https://github.com/symengine/symengine/pull/1704. You are welcome to push to it directly.

**Comment by isuruf** [comment]
Fixed in #1704

**Comment by isuruf** [comment]
We are only exposing a simple `std::string dumps(const Basic &x)` and `RCP<const Basic> loads(std::string &)` interface. Therefore I don't think the global state issue that @bluescarni mentioned will not be there.
